render/render_test.go

```diff
@@ -18,73 +18,150 @@ const (
 name: my-app`
 )
 
-func TestRenderMissingValue(t *testing.T) {
-	configFiles := []composetypes.ConfigFile{
-		{
-			Config: map[string]interface{}{
-				"version": "3",
-				"services": map[string]interface{}{
-					"foo": map[string]interface{}{
-						"image": "${imageName}:${version}",
-					},
-				},
-			},
-		},
+func TestSubstituteBracedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "${front.port}:80"
+`
+	parameters := map[string]string{
+		"front.port": "8080",
 	}
-	finalEnv := map[string]string{
-		"imageName": "foo",
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "8080:80"
+`)
+}
+
+func TestSubstituteNamedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  back:
+    ports:
+     - "$back.port:90"
+`
+	parameters := map[string]string{
+		"back.port": "9000",
 	}
-	_, err := render("foo.dockerapp", configFiles, finalEnv, nil)
-	assert.Check(t, err != nil)
-	assert.Check(t, is.ErrorContains(err, "required variable"))
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  back:
+    ports:
+     - "9000:90"
+`)
 }
 
-func TestRender(t *testing.T) {
-	configFiles := []composetypes.ConfigFile{
-		{
-			Config: map[string]interface{}{
-				"version": "3",
-				"services": map[string]interface{}{
-					"foo": map[string]interface{}{
-						"image":   "busybox:${version}",
-						"command": []interface{}{"-text", "${foo.bar}"},
-					},
-				},
-			},
-		},
+func TestRenderFailOnDefaultParamValueInCompose(t *testing.T) {
```

Comment:
The same test is missing for the `${parameter:?err}' syntax, and `${parameter?err}'.
This test should also check the syntax `${front.port-9090}:80` to be exhaustive.
As @ndeloof commented, regexps are difficult too read and maintain, and are really error prone.
We need to test every features here, otherwise we might miss something in the future while refactoring.